pallets/services-payment/src/lib.rs

```diff
@@ -503,28 +528,71 @@ impl<T: Config> AuthorNotingHook<T::AccountId> for Pallet<T> {
-                ),
-                Ok(imbalance) => {
-                    T::OnChargeForCollatorAssignment::on_unbalanced(imbalance);
+impl<T: Config> CollatorAssignmentHook<BalanceOf<T>> for Pallet<T> {
+    // is_parathread parameter for future use to apply different logic
+    fn on_collators_assigned(
+        para_id: ParaId,
+        maybe_tip: Option<&BalanceOf<T>>,
+        _is_parathread: bool,
+    ) -> Result<Weight, DispatchError> {
+        // Withdraw assignment fee
+        let maybe_assignment_imbalance =
+            if Pallet::<T>::burn_collator_assignment_free_credit_for_para(&para_id).is_err() {
+                let (amount_to_charge, _weight) =
+                    T::ProvideCollatorAssignmentCost::collator_assignment_cost(&para_id);
+                Some(T::Currency::withdraw(
+                    &Self::parachain_tank(para_id),
+                    amount_to_charge,
+                    WithdrawReasons::FEE,
+                    ExistenceRequirement::KeepAlive,
+                )?)
+            } else {
+                None
+            };
+
+        if let Some(&tip) = maybe_tip {
+            // Only charge the tip to the paras that had a max tip set
+            // (aka were willing to tip for being assigned a collator)
+            if MaxTip::<T>::get(para_id).is_some() {
+                match T::Currency::withdraw(
+                    &Self::parachain_tank(para_id),
+                    tip,
+                    WithdrawReasons::TIP,
+                    ExistenceRequirement::KeepAlive,
+                ) {
+                    Err(e) => {
+                        // Return assignment imbalance to tank on error
+                        if let Some(assignment_imbalance) = maybe_assignment_imbalance {
+                            T::Currency::resolve_creating(
+                                &Self::parachain_tank(para_id),
+                                assignment_imbalance,
+                            );
+                        }
+                        return Err(e);
+                    }
+                    Ok(tip_imbalance) => {
+                        Self::deposit_event(Event::<T>::CollatorAssignmentTipCollected {
+                            para_id,
+                            payer: Self::parachain_tank(para_id),
+                            tip,
+                        });
+                        T::OnChargeForCollatorAssignmentTip::on_unbalanced(tip_imbalance);
+                    }
                 }
             }
         }
-        T::WeightInfo::on_collators_assigned()
+
+        if let Some(assignment_imbalance) = maybe_assignment_imbalance {
+            T::OnChargeForCollatorAssignment::on_unbalanced(assignment_imbalance);
+        }
+
+        Ok(T::WeightInfo::on_collators_assigned())
+    }
+}
+
+impl<T: Config> CollatorAssignmentTip<BalanceOf<T>> for Pallet<T> {
+    fn get_para_tip(para_id: ParaId) -> Option<BalanceOf<T>> {
+        MaxTip::<T>::get(para_id)
```

Comment:
In this file the `para_deregistered` hook should clean all the storage items. It's missing `MaxTip` and also `MaxCorePrice` because I forgot to add it

Reply:
Added both